### PR TITLE
week 3: Authentication

### PR DESCRIPTION
**Pull request by pr3pony**

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -27,18 +27,16 @@
...
 import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
 
 /** Servlet that returns comments data.*/
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  private DatastoreService mDatastore;
+  private DatastoreService mDatastore = DatastoreServiceFactory.getDatastoreService();
```
Comment (shadyof) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: if we decide to move it out from #init, then we can consider add 'final' for it

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -27,18 +27,16 @@
...
   private final ArrayList<String> mComments = new ArrayList<String>();;
```
Comment (shadyof) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Ah seems I didn't notice here last time, there are two ;;

File: portfolio/src/main/java/com/google/sps/servlets/UserServlet.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+/** Servlet that returns user info.*/
+@WebServlet("/user")
+public class UserServlet extends HttpServlet {
+  private class UserStatus { 
+    public boolean isUserLoggedIn;
+    public String userEmail;
+    public String logoutUrl;
+    public String loginUrl;
+    public UserStatus() {
```
Comment (shadyof) [inline portfolio/src/main/java/com/google/sps/servlets/UserServlet.java]: consider the access modifier, do they need to be public?

File: portfolio/src/main/java/com/google/sps/servlets/UserServlet.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        isUserLoggedIn = false;
+        userEmail = null;
+        loginUrl = null;
+        logoutUrl = null;
```
Comment (shadyof) [inline portfolio/src/main/java/com/google/sps/servlets/UserServlet.java]: nit: above assignment for those members seems unnecessary

File: portfolio/src/main/java/com/google/sps/servlets/UserServlet.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+
+/** Servlet that returns user info.*/
+@WebServlet("/user")
+public class UserServlet extends HttpServlet {
+  private class UserStatus {
```
Comment (shadyof) [inline portfolio/src/main/java/com/google/sps/servlets/UserServlet.java]: could this class be 'final' and 'static' ?

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -40,14 +40,28 @@ function addRandomGreeting() {
...
+    fetch('/user').then(response => response.json()).then(userStatus => {
+        const userContainer = document.getElementById('user-container');
+        if (userStatus.isUserLoggedIn) {
+            userContainer.innerHTML = '<span style="float:left;">username: ' + userStatus.userEmail + '</span>'
+                + '<span style="float:right;"><a href="' + userStatus.logoutUrl + '">logout</a></span>';
+        }
+        else {
+            userContainer.innerHTML = '<span style="float:left;">username: None</span>'
+                + '<span style="float:right;"><a href="' + userStatus.loginUrl + '">login</a></span>';
+        }
```
Comment (shadyof) [inline portfolio/src/main/webapp/script.js]: just question so doesn't need to modify: I guess cloud shell doesn't provide any hint or suggestion about line length?

File: portfolio/src/main/webapp/script.js [hunk 7]
```diff
@@ -40,14 +40,28 @@ function addRandomGreeting() {
...
+    });
+    if (window.location.search == '?login') {
+        alert("Pleas login to leave comments.");
```
Comment (shadyof) [inline portfolio/src/main/webapp/script.js]: typo? "please"

**Comment by pr3pony** [comment]
Is it good now?

**Review by shadyof — APPROVED** [review]
Sorry to reply so late 
LGTM!

**Pull request merged**